src/libsyntax/ast.rs

```diff
@@ -1746,13 +1730,21 @@ impl StructFieldKind {
     }
 }
 
+#[derive(Clone, Copy, PartialEq, Eq, RustcEncodable, RustcDecodable, Hash, Debug)]
+pub enum VariantKind {
+    Dict,
```

Comment:
Hmm, I guess using StructField is better than Named/UnnamedField then.

Could you have a function which takes a per-field closure and applies it to each field in Struct/Tuple but does nothing in Unit case? That seems like a better solution than a field getter (or return an empty slice in the Unit case (not sure if that is possible without allocating, but it ought to be). I definitely don't like keeping fields for DataKind::Unit.

Note that having the id here helps with the ctor_id problem too.